pkg/streamingpromql/operators/aggregations/quantile.go

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package aggregations
+
+import (
+	"context"
+	"math"
+	"unsafe"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+// QuantileAggregation is a small wrapper around Aggregation to pre-process and validate
+// the quantile parameter and fill it into Aggregation.ParamData
+type QuantileAggregation struct {
+	Param                    types.ScalarOperator
+	Aggregation              *Aggregation
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	Annotations              *annotations.Annotations
+}
+
+func NewQuantileAggregation(
+	inner types.InstantVectorOperator,
+	param types.ScalarOperator,
+	timeRange types.QueryTimeRange,
+	grouping []string,
+	without bool,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+) (*QuantileAggregation, error) {
+
+	a, err := NewAggregation(
+		inner,
+		timeRange,
+		grouping,
+		without,
+		parser.QUANTILE,
+		memoryConsumptionTracker,
+		annotations,
+		expressionPosition,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	q := &QuantileAggregation{
+		Aggregation:              a,
+		Param:                    param,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		Annotations:              annotations,
+	}
+
+	return q, nil
+}
+
+func (q *QuantileAggregation) ExpressionPosition() posrange.PositionRange {
+	return q.Aggregation.ExpressionPosition()
+}
+
+func (q *QuantileAggregation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	var err error
+	q.Aggregation.ParamData, err = q.Param.GetValues(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Validate the parameter now so we only have to do it once for each group
+	for _, p := range q.Aggregation.ParamData.Samples {
+		if math.IsNaN(p.F) || p.F < 0 || p.F > 1 {
+			q.Annotations.Add(annotations.NewInvalidQuantileWarning(p.F, q.Param.ExpressionPosition()))
+		}
+	}
+
+	return q.Aggregation.SeriesMetadata(ctx)
+}
+
+func (q *QuantileAggregation) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	return q.Aggregation.NextSeries(ctx)
+}
+
+func (q *QuantileAggregation) Close() {
+	if q.Aggregation.ParamData.Samples != nil {
```

Comment:
It feels weird that we're reaching into `q.Aggregation` here - I think it'd be OK to move this into `Aggregation.Close()`.